File: vpbuf/tests/maps/cpp/src/maps.cc
// Software released under the MIT license (see project root for license file)

#include <cstddef>
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <cassert>
#include <typeindex> // needed for class polymorphism

using namespace std;

typedef char byte;
typedef vector<byte> ByteVec;

// buffer management helper classes and functions
#include "buffer_context.cc"


// include classes to be serialized
#include "persist.cc"

// include output from vpc code generator
#include "vp_maps.cpp"
using namespace vp_maps;


// specify output directory and file extension for test output

const string out_dir = "./out/";
const string file_ext = ".dat";

// -----------------------------------------------------------------------------
// -----------------------------------------------------------------------------

int main(int argc, char* argv[])
{
   // ------------------------------------------------------------------
   // maps_A -

   {
      string test_id_str = "maps_A";
      int count = 10;
      int base = 50000;

      Header h1;
      OuterA o1;

      {
         // initialize test data structures and write to file
         write_context wc;

         h1.version = -99;
         h1.test_name = test_id_str;


         for (auto i = 0; i < count; i++) {
            auto a = new A;
            a->i1 = base + i;
            a->s1 = "A-" + std::to_string(base + i);
            o1.lookup[base + i] = a;
         }

         write_Header(1, wc, h1);
         write_OuterA(1, wc, o1);

         size_t bytes_out = wc.write_file(out_dir + test_id_str + file_ext);
         std::cout << test_id_str <<" "<< bytes_out <<" bytes written\n";
      }

      Header h2;
      OuterA o2;

      {
         // read the file just written
         read_context rc(out_dir + test_id_str + file_ext);

         read_Header(1, rc, h2);
         read_OuterA(1, rc, o2);

         std::cout << test_id_str <<" "<< rc.get_buffer_size() <<" bytes read\n";
      }

      // compare data written against data read in

      assert(h1.version == h2.version);
      assert(h1.test_name == h2.test_name);
      assert(o1.lookup.size() == o2.lookup.size());
      for (auto ii = o1.lookup.begin();ii != o1.lookup.end(); ii++) {
         auto p = o2.lookup[ii->first];
         assert((ii->second)->s1 == p->s1);
         assert((ii->second)->i1 == p->i1);
      }
   }

   // ------------------------------------------------------------------
   // maps_B -

   {
      string test_id_str = "maps_B";
      int count = 10;
      int base = 50000;

      Header h1;
      OuterB o1;

      {
         // initialize test data structures and write to file
         write_context wc;

         h1.version = -99;
         h1.test_name = test_id_str;


         for (auto i = 0; i < count; i++) {
            auto a = new A;
            a->i1 = base + i;
            a->s1 = "B-" + std::to_string(base + i);
            o1.lookup[to_string(base + i)] = a;
         }

         write_Header(1, wc, h1);
         write_OuterB(1, wc, o1);

         size_t bytes_out = wc.write_file(out_dir + test_id_str + file_ext);
         std::cout << test_id_str <<" "<< bytes_out <<" bytes written\n";
      }

      Header h2;
      OuterB o2;

      {
         // read the file just written
         read_context rc(out_dir + test_id_str + file_ext);

         read_Header(1, rc, h2);
         read_OuterB(1, rc, o2);

         std::cout << test_id_str <<" "<< rc.get_buffer_size() <<" bytes read\n";
      }

      // compare data written against data read in

      assert(h1.version == h2.version);
      assert(h1.test_name == h2.test_name);
      assert(o1.lookup.size() == o2.lookup.size());
      for (auto ii = o1.lookup.begin();ii != o1.lookup.end(); ii++) {
         auto p = o2.lookup[ii->first];
         assert((ii->second)->s1 == p->s1);
         assert((ii->second)->i1 == p->i1);
      }
   }

// ------------------------------------------------------------------
   // maps_C -

   {
      string test_id_str = "maps_C";
      int count = 10;
      int base = 50000;

      Header h1;
      OuterC o1;

      {
         // initialize test data structures and write to file
         write_context wc;

         h1.version = -99;
         h1.test_name = test_id_str;


         for (auto i = 0; i < count; i++) {
            o1.lookup[base + i] = "C-" + to_string(base + i);
         }

         write_Header(1, wc, h1);
         write_OuterC(1, wc, o1);

         size_t bytes_out = wc.write_file(out_dir + test_id_str + file_ext);
         std::cout << test_id_str <<" "<< bytes_out <<" bytes written\n";
      }

      Header h2;
      OuterC o2;

      {
         // read the file just written
         read_context rc(out_dir + test_id_str + file_ext);

         read_Header(1, rc, h2);
         read_OuterC(1, rc, o2);

         std::cout << test_id_str <<" "<< rc.get_buffer_size() <<" bytes read\n";
      }

      // compare data written against data read in

      assert(h1.version == h2.version);
      assert(h1.test_name == h2.test_name);
      assert(o1.lookup.size() == o2.lookup.size());
      for (auto ii = o1.lookup.begin();ii != o1.lookup.end(); ii++) {
         auto p = o2.lookup[ii->first];
         assert((ii->second) == p);
      }
   }

   // ------------------------------------------------------------------
   // maps_D -

   {
      string test_id_str = "maps_D";
      int count = 10;
      int base = 50000;

      Header h1;
      OuterD o1;

      {
         // initialize test data structures and write to file
         write_context wc;

         h1.version = -99;
         h1.test_name = test_id_str;


         for (auto i = 0; i < count; i++) {
            int j = base + i;
            o1.lookup[to_string(j)] = new D1(j, "D1-" + to_string(j));
         }

         write_Header(1, wc, h1);
         write_OuterD(1, wc, o1);

         size_t bytes_out = wc.write_file(out_dir + test_id_str + file_ext);
         std::cout << test_id_str <<" "<< bytes_out <<" bytes written\n";
      }

      Header h2;
      OuterD o2;

      {
         // read the file just written
         read_context rc(out_dir + test_id_str + file_ext);

         read_Header(1, rc, h2);
         read_OuterD(1, rc, o2);

         std::cout << test_id_str <<" "<< rc.get_buffer_size() <<" bytes read\n";
      }

      // compare data written against data read in

      assert(h1.version == h2.version);
      assert(h1.test_name == h2.test_name);
      assert(o1.lookup.size() == o2.lookup.size());
      for (auto ii = o1.lookup.begin();ii != o1.lookup.end(); ii++) {
         auto p = o2.lookup[ii->first];
         assert((ii->second)->s == p->s);
         assert((ii->second)->i == p->i);
      }
   }

   // ------------------------------------------------------------------

}

// -----------------------------------------------------------------------------

File: vpbuf/tests/maps/python/maps.py
#!/usr/bin/python3

# Software released under the MIT license (see project root for license file)

from pyamf import amf3
from vp_maps.persist import *
from vp_maps.vp_maps import *

# ------------------------------------------------------------------------------

def buffer_to_file(file_name, encoder):
    output_buffer = amf3.DataOutput(encoder)
    f = open(file_name, "wb")
    bytes = output_buffer.stream.getvalue()
    f.write(bytes)
    f.close()
    return len(bytes)

def file_to_buffer(file_name):
    f_in = open(file_name, "rb")
    istream = amf3.util.BufferedByteStream(f_in)
    f_in.close()
    return istream

out_dir = "./out/"
file_ext = ".dat"

# ------------------------------------------------------------------------------

class write_context:
    def __init__(self, ver = 1):
        self.stream = amf3.util.BufferedByteStream()
        self.encoder = amf3.Encoder(self.stream)
        self.encoder.string_references = False # disables string caching

        self.reorder_map = {}
        self.ver = ver
        self.set_version(ver)

        self.salt_map = {}
        self.salt_map["SaltShaker"] = SaltShaker(ver)
        self.salt_map["PepperShaker"] = PepperShaker(ver)

    def set_version(self, ver):
        self.ver = ver
        init_reorder_map(self.reorder_map, ver, 1)

class read_context:
    def __init__(self, test_name, ver = 1):
        self.istream = file_to_buffer(out_dir + test_name + file_ext)
        self.decoder = amf3.Decoder(self.istream)
        self.bytes_read = len(self.istream)
        print(test_name, len(self.istream), 'bytes read')
        self.reorder_map = {}
        self.ver = ver
        self.set_version(ver)

    def set_version(self, ver):
        self.ver = ver
        init_reorder_map(self.reorder_map, ver, 1)

# ------------------------------------------------------------------------------
# maps_A -

class Test_maps_A:
    def __init__(self):
        self.test_name = "maps_A"
        self.count = 10
        self.base = 50000

        self.h1 = Header()  # initialized and then serialized
        self.o1 = OuterA()

        self.h2 = Header()  # deserialized and validated
        self.o2 = OuterA()

        self.bytes_read = 0
        self.bytes_written = 0

    # initialize the data structures under test and serialize

    def serialize(self):
        wc = write_context()

        self.h1.version = -99
        self.h1.test_name = self.test_name

        for i in range(0, self.count):
            j = self.base + i
            self.o1.lookup[j] = A(j, 'A-' + str(j))

        write_Header(wc, self.h1)
        write_OuterA(wc, self.o1)

        out_file = out_dir + self.test_name + file_ext
        self.bytes_written = buffer_to_file(out_file, wc.encoder)

        print(self.test_name, self.bytes_written, 'bytes written')

    # deserialize the data structures under test

    def load(self):
        rc = read_context(self.test_name)

        self.h2 = read_Header(rc)
        self.o2 = read_OuterA(rc)

        self.bytes_read = rc.bytes_read

    # compare the serialized and deserialized data structures against each other

    def validate(self):
        assert(self.h1.version == self.h2.version)
        assert(self.h1.test_name == self.h2.test_name)
        assert(self.bytes_read == self.bytes_written)

        assert(len(self.o1.lookup) == len(self.o2.lookup))

        for k in self.o1.lookup.keys():
            assert(self.o1.lookup[k].i1 == self.o2.lookup[k].i1)
            assert(self.o1.lookup[k].s1 == self.o2.lookup[k].s1)

a = Test_maps_A()
a.serialize()
a.load()
a.validate()

# ------------------------------------------------------------------------------
# maps_B -

class Test_maps_B:
    def __init__(self):
        self.test_name = "maps_B"
        self.count = 10
        self.base = 50000

        self.h1 = Header()  # initialized and then serialized
        self.o1 = OuterB()

        self.h2 = Header()  # deserialized and validated
        self.o2 = OuterB()

        self.bytes_read = 0
        self.bytes_written = 0

    # initialize the data structures under test and serialize

    def serialize(self):
        wc = write_context()

        self.h1.version = -99
        self.h1.test_name = self.test_name

        for i in range(0, self.count):
            j = self.base + i
            self.o1.lookup[str(j)] = A(j, 'B-' + str(j))

        write_Header(wc, self.h1)
        write_OuterB(wc, self.o1)

        out_file = out_dir + self.test_name + file_ext
        self.bytes_written = buffer_to_file(out_file, wc.encoder)

        print(self.test_name, self.bytes_written, 'bytes written')

    # deserialize the data structures under test

    def load(self):
        rc = read_context(self.test_name)

        self.h2 = read_Header(rc)
        self.o2 = read_OuterB(rc)

        self.bytes_read = rc.bytes_read

    # compare the serialized and deserialized data structures against each other

    def validate(self):
        assert(self.h1.version == self.h2.version)
        assert(self.h1.test_name == self.h2.test_name)
        assert(self.bytes_read == self.bytes_written)

        assert(len(self.o1.lookup) == len(self.o2.lookup))

        for k in self.o1.lookup.keys():
            assert(self.o1.lookup[k].i1 == self.o2.lookup[k].i1)
            assert(self.o1.lookup[k].s1 == self.o2.lookup[k].s1)

a = Test_maps_B()
a.serialize()
a.load()
a.validate()

# ------------------------------------------------------------------------------
# maps_C -

class Test_maps_C:
    def __init__(self):
        self.test_name = "maps_C"
        self.count = 10
        self.base = 50000

        self.h1 = Header()  # initialized and then serialized
        self.o1 = OuterC()

        self.h2 = Header()  # deserialized and validated
        self.o2 = OuterC()

        self.bytes_read = 0
        self.bytes_written = 0

    # initialize the data structures under test and serialize

    def serialize(self):
        wc = write_context()

        self.h1.version = -99
        self.h1.test_name = self.test_name

        for i in range(0, self.count):
            j = self.base + i
            self.o1.lookup[j] = 'C-' + str(j)

        write_Header(wc, self.h1)
        write_OuterC(wc, self.o1)

        out_file = out_dir + self.test_name + file_ext
        self.bytes_written = buffer_to_file(out_file, wc.encoder)

        print(self.test_name, self.bytes_written, 'bytes written')

    # deserialize the data structures under test

    def load(self):
        rc = read_context(self.test_name)

        self.h2 = read_Header(rc)
        self.o2 = read_OuterC(rc)

        self.bytes_read = rc.bytes_read

    # compare the serialized and deserialized data structures against each other

    def validate(self):
        assert(self.h1.version == self.h2.version)
        assert(self.h1.test_name == self.h2.test_name)
        assert(self.bytes_read == self.bytes_written)
        assert(len(self.o1.lookup) == len(self.o2.lookup))

        for k in self.o1.lookup.keys():
            assert(self.o1.lookup[k] == self.o2.lookup[k])


a = Test_maps_C()
a.serialize()
a.load()
a.validate()

# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# maps_D -

class Test_maps_D:
    def __init__(self):
        self.test_name = "maps_D"
        self.count = 10
        self.base = 50000

        self.h1 = Header()  # initialized and then serialized
        self.o1 = OuterD()

        self.h2 = Header()  # deserialized and validated
        self.o2 = OuterD()

        self.bytes_read = 0
        self.bytes_written = 0

    # initialize the data structures under test and serialize

    def serialize(self):
        wc = write_context()

        self.h1.version = -99
        self.h1.test_name = self.test_name

        for i in range(0, self.count):
            j = self.base + i
            self.o1.lookup[str(j)] = D1(j, 'D1-' + str(j))

        write_Header(wc, self.h1)
        write_OuterD(wc, self.o1)

        out_file = out_dir + self.test_name + file_ext
        self.bytes_written = buffer_to_file(out_file, wc.encoder)

        print(self.test_name, self.bytes_written, 'bytes written')

    # deserialize the data structures under test

    def load(self):
        rc = read_context(self.test_name)

        self.h2 = read_Header(rc)
        self.o2 = read_OuterD(rc)

        self.bytes_read = rc.bytes_read

    # compare the serialized and deserialized data structures against each other

    def validate(self):
        assert(self.h1.version == self.h2.version)
        assert(self.h1.test_name == self.h2.test_name)
        assert(self.bytes_read == self.bytes_written)
        assert(len(self.o1.lookup) == len(self.o2.lookup))

        for k in self.o1.lookup.keys():
            assert(self.o1.lookup[k].i == self.o2.lookup[k].i)
            assert(self.o1.lookup[k].s == self.o2.lookup[k].s)

            assert(self.o1.lookup[k].apod.i1 == self.o2.lookup[k].apod.i1)
            assert(self.o1.lookup[k].apod.s1 == self.o2.lookup[k].apod.s1)

            assert(self.o1.lookup[k].aref.i1 == self.o2.lookup[k].aref.i1)
            assert(self.o1.lookup[k].aref.s1 == self.o2.lookup[k].aref.s1)

a = Test_maps_D()
a.serialize()
a.load()
a.validate()

# ------------------------------------------------------------------------------

# ------------------------------------------------------------------------------
# maps_E -

class Test_maps_E:
    def __init__(self):
        self.test_name = "maps_E"
        self.count = 10
        self.base = 50000

        self.h1 = Header()  # initialized and then serialized
        self.o1 = Omlette()

        self.h2 = Header()  # deserialized and validated
        self.o2 = Omlette()

        self.bytes_read = 0
        self.bytes_written = 0

    # initialize the data structures under test and serialize

    def serialize(self):
        wc = write_context()

        self.h1.version = -99
        self.h1.test_name = self.test_name

        for i in range(0, self.count):
            j = self.base + i
            self.o1.lookup[j] = Egg(j, 'EGG-' + str(j))

        write_Header(wc, self.h1)
        write_Omlette(wc, self.o1)

        out_file = out_dir + self.test_name + file_ext
        self.bytes_written = buffer_to_file(out_file, wc.encoder)

        print(self.test_name, self.bytes_written, 'bytes written')

    # deserialize the data structures under test

    def load(self):
        rc = read_context(self.test_name)

        self.h2 = read_Header(rc)
        self.o2 = read_Omlette(rc)

        self.bytes_read = rc.bytes_read

    # compare the serialized and deserialized data structures against each other

    def validate(self):
        assert(self.h1.version == self.h2.version)
        assert(self.h1.test_name == self.h2.test_name)
        assert(self.bytes_read == self.bytes_written)

        assert(len(self.o1.lookup) == len(self.o2.lookup))

        for k in self.o1.lookup.keys():
            assert(self.o1.lookup[k].i1 == self.o2.lookup[k].i1)
            assert(self.o1.lookup[k].s1 == self.o2.lookup[k].s1)

e = Test_maps_E()
e.serialize()
e.load()
e.validate()
